pkg/chartverifier/certificate.go

```diff
@@ -16,76 +16,88 @@
 }
 
-func newMetadata(name, version, chartUri, toolVersion string) *metadata {
-	return &metadata{
-		RunMetadata: runMetadata{
-			ChartUri: chartUri,
-			Version:  toolVersion,
-		},
-		ChartMetadata: chartMetadata{
-			Name:    name,
-			Version: version,
-		},
-	}
+type ToolMetadata struct {
+	Version                    string `json:"verifier-version" yaml:"verifier-version"`
+	ChartUri                   string `json:"chart-uri" yaml:"chart-uri"`
+	Digest                     string `json:"digest" yaml:"digest"`
+	LastCertifiedTime          string `json:"lastCertifiedTime" yaml:"lastCertifiedTime"`
+	CertifiedOpenShiftVersions string `json:"certifiedOpenShiftVersions" yaml:"certifiedOpenShiftVersions"`
 }
 
-type certificate struct {
-	Ok             bool           `json:"ok" yaml:"ok"`
-	Metadata       *metadata      `json:"metadata" yaml:"metadata"`
-	CheckResultMap checkResultMap `json:"results" yaml:"results"`
+type CheckReport struct {
+	Check   string           `json:"check" yaml:"check"`
+	Type    checks.CheckType `json:"type" yaml:"type"`
+	Outcome OutcomeType      `json:"outcome" yaml:"outcome"`
+	Reason  string           `json:"reason" yaml:"reason"`
 }
 
-type checkResultMap map[string]checkResult
+func newCertificate() Certificate {
 
-type checkResult struct {
-	Ok     bool   `json:"ok" yaml:"ok"`
-	Reason string `json:"reason" yaml:"reason"`
-}
+	certificate := Certificate{Apiversion: CertificateApiVersion, Kind: CertificateKind}
+	certificate.Metadata = CertificateMetadata{}
+	certificate.Metadata.ToolMetadata = ToolMetadata{}
 
-func newCertificate(name, version, chartUri, toolVersion string, ok bool, resultMap checkResultMap) Certificate {
-	return &certificate{
-		Metadata:       newMetadata(name, version, chartUri, toolVersion),
-		Ok:             ok,
-		CheckResultMap: resultMap,
-	}
+	return certificate
 }
 
-func (c *certificate) IsOk() bool {
-	return c.Ok
+func (c *Certificate) AddCheck(checkName string, checkType checks.CheckType) *CheckReport {
+	newCheck := CheckReport{}
+	newCheck.Check = checkName
+	newCheck.Type = checkType
+	newCheck.Outcome = PassOutcomeType
+	c.Results = append(c.Results, &newCheck)
+	return &newCheck
 }
 
-func (c *certificate) String() string {
-	report := "Tool:\n" +
-		"  verifier-version: " + c.Metadata.RunMetadata.Version + "\n" +
-		"  chart-uri: " + c.Metadata.RunMetadata.ChartUri + "\n" +
-		"Chart:\n" +
-		"  Name: " + c.Metadata.ChartMetadata.Name + "\n" +
-		"  version: " + c.Metadata.ChartMetadata.Version + "\n" +
-		"ok: " + strconv.FormatBool(c.Ok) + "\n" +
-		"\n"
-
-	for k, v := range c.CheckResultMap {
-		report += k + ":\n" +
-			"\tok: " + strconv.FormatBool(v.Ok) + "\n" +
-			"\treason: " + v.Reason + "\n"
+func (cr *CheckReport) SetResult(outcome bool, reason string) {
+	if outcome {
+		cr.Outcome = PassOutcomeType
+	} else {
+		cr.Outcome = FailOutcomeType
 	}
+	cr.Reason = reason
+}
 
-	return report
+func (c *Certificate) IsOk() bool {
```

Comment:
did not we decide that the tool is not going to aggregate/reason about the checks, i.e. what are mandatory and what not?

Reply:
Yes, I did this as a helper function for the tests.

Reply:
can you move it then under `certificate_test.go` file, or at least make it private?

Reply:
I would ask to defer on this to #78 also. The method was not introduced by this PR and is the only method defined in a Certificate interface. To remove the interface creates knock on effects in other places and I can clean it up in #78.